apps/laboratory/src/hooks/usePermissions.ts

```diff
@@ -0,0 +1,191 @@
+import {
+  ENTRYPOINT_ADDRESS_V07,
+  createBundlerClient,
+  getPackedUserOperation,
+  getUserOperationHash
+} from 'permissionless'
+import { pimlicoBundlerActions, pimlicoPaymasterActions } from 'permissionless/actions/pimlico'
+import { type UserOperation } from 'permissionless/types'
+import { createPublicClient, http, signatureToHex, type PublicClient } from 'viem'
+import { foundry, sepolia, type Chain } from 'wagmi/chains'
+import { sign } from 'viem/accounts'
+import { useUserOpBuilder, type Execution } from './useUserOpBuilder'
+import { bigIntReplacer } from '../utils/CommonUtils'
+
+export function usePermissions() {
+  const { getCallDataWithContext, getNonceWithContext, getSignatureWithContext } =
+    useUserOpBuilder()
+
+  async function prepareUserOperationWithPermissions(
+    publicClient: PublicClient,
+    args: {
+      actions: Execution[]
+      permissions: GrantPermissionsReturnType
+    }
+  ): Promise<UserOperation<'v0.7'>> {
+    const { permissions, actions } = args
+    const { factory, factoryData, signerData, permissionsContext } = permissions
+
+    if (!signerData?.userOpBuilder || !signerData.submitToAddress || !permissionsContext) {
+      throw new Error(`Invalid permissions ${JSON.stringify(permissions, bigIntReplacer)}`)
+    }
+
+    const nonce = await getNonceWithContext(publicClient, {
+      userOpBuilderAddress: signerData.userOpBuilder,
+      sender: signerData.submitToAddress,
+      permissionsContext: permissionsContext as `0x${string}`
+    })
+
+    const callData = await getCallDataWithContext(publicClient, {
+      userOpBuilderAddress: signerData.userOpBuilder,
+      sender: signerData.submitToAddress,
+      permissionsContext: permissionsContext as `0x${string}`,
+      actions
+    })
+
+    const userOp: UserOperation<'v0.7'> = {
+      sender: signerData.submitToAddress,
+      factory,
+      factoryData: factoryData ? (factoryData as `0x${string}`) : undefined,
+      nonce,
+      callData,
+      callGasLimit: BigInt(2000000),
+      verificationGasLimit: BigInt(2000000),
+      preVerificationGas: BigInt(2000000),
+      maxFeePerGas: BigInt(0),
+      maxPriorityFeePerGas: BigInt(0),
+      signature: '0x'
+    }
+
+    return userOp
+  }
+
+  async function signUserOperationWithECDSAKeyAndPermissions(
+    publicClient: PublicClient,
+    args: {
+      ecdsaPrivateKey: `0x${string}`
+      userOp: UserOperation<'v0.7'>
+      permissions: GrantPermissionsReturnType
+      chain: Chain
+    }
+  ): Promise<`0x${string}`> {
+    const { ecdsaPrivateKey, userOp, chain, permissions } = args
+
+    const { signerData, permissionsContext } = permissions
+
+    if (!signerData?.userOpBuilder || !signerData.submitToAddress || !permissionsContext) {
+      throw new Error(`Invalid permissions ${JSON.stringify(permissions, bigIntReplacer)}`)
+    }
+    const userOpHash = getUserOperationHash({
+      userOperation: {
+        ...userOp
+      },
```

Comment:
If we don't pass any other fields here, we don't have to destruct the object and directly use the ref. (In case it's not required for cloning object purposes)
```suggestion
    const userOpHash = getUserOperationHash({
      userOperation: userOp,
```